// cirrus-ping-auth-module/dist/logger.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _chalk = _interopRequireDefault(require("chalk"));

var loggingEnabled = process.env.ENABLE_AUTH_LOGGING === 'true';

if (!loggingEnabled) {
  console.log("".concat(_chalk["default"].bgWhite.bold('[AUTH]'), " : Logs are disabled."));
}
/**
 * If logs are enabled in the .env file - we log
 * We log in a way that the user knows that it's the auth module logging.
 * @param  {String} string what should be printed
 */


var prefixPrint = function prefixPrint(string) {
  if (loggingEnabled) {
    console.log("".concat(_chalk["default"].bgWhite.bold('[AUTH]'), " ").concat(string));
  }
};

var _default = {
  error: function error(message) {
    return prefixPrint(_chalk["default"].bgRed(message));
  },
  warn: function warn(message) {
    return prefixPrint(_chalk["default"].bgRed(message));
  },
  success: function success(message) {
    return prefixPrint(_chalk["default"].bgGreen(message));
  },
  info: function info(message) {
    return prefixPrint(_chalk["default"].bgCyan(message));
  },
  plain: function plain(message) {
    return prefixPrint(message);
  }
};
exports["default"] = _default;
