// cirrus-oidc-auth-module/dist/logger.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.plain = exports.info = exports.success = exports.warn = exports.error = void 0;

var _chalk = _interopRequireDefault(require("chalk"));

/* eslint no-console: 0 */
var ENABLE_AUTH_LOGGING = process.env.ENABLE_AUTH_LOGGING === 'true';

if (!ENABLE_AUTH_LOGGING) {
  console.log("".concat(_chalk["default"].bgWhite.bold('[AUTH]'), " : Logs are disabled."));
}
/**
 * If logs are enabled in the .env file - we log
 * We log in a way that the user knows that it's the auth module logging.
 * @param  {String} string what should be printed
 */


var print = function print(string) {
  if (ENABLE_AUTH_LOGGING) {
    console.log("".concat(_chalk["default"].bgWhite.bold('[AUTH]'), " ").concat(string));
  }
};

var error = function error(message) {
  return print(_chalk["default"].bgRed(message));
};

exports.error = error;

var warn = function warn(message) {
  return print(_chalk["default"].bgRed(message));
};

exports.warn = warn;

var success = function success(message) {
  return print(_chalk["default"].bgGreen(message));
};

exports.success = success;

var info = function info(message) {
  return print(_chalk["default"].bgCyan(message));
};

exports.info = info;

var plain = function plain(message) {
  return print(message);
};

exports.plain = plain;
